### ISSUE TITLE
[REQUEST] Autocast for BF16

### ISSUE THREAD

**Issue by MrCrHaM** [microsoft/DeepSpeed#4772]
**Is your feature request related to a problem? Please describe.**
Currently for FP16 training, there is an option for autocast, while casting must be done manually for BF16 training, which is something like the following,
```python
if model_engine.bfloat16_enabled():
    data_batch = data_batch.bfloat16()
```

**Describe the solution you'd like**
An option for autocasting BF16 just like the FP16 counterpart would be great. Or is there any particular reason for not implementing this in the first place?

**Comment by ethansmith2000** [comment]
If i have it right, it seems that bf16 will perform the entire training with all weights and everything at bf16 rather than mixed. I see the AMP config option notes that it doesn't support being used with Zero, maybe thats why?

**Comment by MrCrHaM** [comment]
> If i have it right, it seems that bf16 will perform the entire training with all weights and everything at bf16 rather than mixed.

I don't think so. By looking at the implementation, it seems that the model updates are calculated in FP32 just like a regular mixed precision fashion.
https://github.com/microsoft/DeepSpeed/blob/ee7db48373bffedba9e5f5b570b7e07202b7a875/deepspeed/runtime/bf16_optimizer.py#L30

**Comment by ethansmith2000** [comment]
yes but the weights themselves and the optimizer states are held in bf16, that bit is new to me
